packages/superset-ui-core/src/SupersetClient.js

```diff
@@ -0,0 +1,180 @@
+import callApi from './callApi';
+
+const AUTH_QUEUE_POLL_MS = 30;
+
+class SupersetClient {
+  constructor(config) {
+    const {
+      protocol = 'http',
+      host = '',
+      headers = {},
+      mode = 'same-origin',
+      timeout,
+      credentials,
+    } = config;
+
+    this.headers = headers;
+    this.host = host;
+    this.mode = mode;
+    this.timeout = timeout;
+    this.protocol = protocol;
+    this.credentials = credentials;
+    this.csrfToken = null;
+    this.didAuthSuccessfully = false;
+    this.requestingCsrf = false;
+
+    this.getUnauthorizedError = this.getUnauthorizedError.bind(this);
+    this.waitForCSRF = this.waitForCSRF.bind(this);
+  }
+
+  isAuthenticated() {
+    return this.didAuthSuccessfully;
+  }
+
+  init() {
+    return this.getCSRFToken();
+  }
+
+  getCSRFToken() {
+    this.requestingCsrf = true;
```

Comment:
Instead of having boolean flags `this.requestingCsrf`, we can store the promise returned from `callApi()` in `this.requestingCsrf`. The initial value of `this.requestingCsrf` can be `Promise.reject('not initialized')`. 

This may remove the need for `this.waitForCSRF` and polling/setTimeout mechanism as `this.ensureAuth` can just `return this.requestingCsrf;` directly.

Reply:
that's a great idea, I'll try to move toward that 👍